From d8c2cff51c23dc387e96846bf7b5b6626e9048cf Mon Sep 17 00:00:00 2001
From: Daniel Drubin <idaniel.drubin@intel.com>
Date: Tue, 9 Feb 2016 18:27:17 +0200
Subject: [PATCH 06/12] ISH: fix for FW reset flow

Fixed releasing/stopping clients in HECI user-mode API
provider and HID client driver. Removed extraneous
locking in release client path

Change-Id: I19b010255ed9c0a4e6774fabf1adfa5ff244ca8a
Signed-off-by: Daniel Drubin <daniel.drubin@intel.com>
Signed-off-by: Kishore Kadiyala <kishore.kadiyala@intel.com>
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-19443
Reviewed-on: https://android.intel.com:443/468133
---
 drivers/misc/intel-ish/heci-api.c        | 8 ++------
 drivers/misc/intel-ish/heci-hid-client.c | 7 +++++++
 drivers/misc/intel-ish/platform-config.h | 2 +-
 3 files changed, 10 insertions(+), 7 deletions(-)

diff --git a/drivers/misc/intel-ish/heci-api.c b/drivers/misc/intel-ish/heci-api.c
index a056202..19dd142 100644
--- a/drivers/misc/intel-ish/heci-api.c
+++ b/drivers/misc/intel-ish/heci-api.c
@@ -124,7 +124,6 @@ static int heci_release(struct inode *inode, struct file *file)
 	struct heci_cl *cl = file->private_data;
 	struct heci_device *dev;
 	int rets = 0;
-	unsigned int flags;
 
 	ISH_DBG_PRINT(KERN_ALERT "%s(): +++\n", __func__);
 	if (WARN_ON(!cl || !cl->dev))
@@ -136,10 +135,9 @@ static int heci_release(struct inode *inode, struct file *file)
 	 * May happen if device sent FW reset or was intentionally
 	 * halted by host SW. The client is then invalid
 	 */
-	if (dev->dev_state != HECI_DEV_ENABLED)
-		return	0;
 
-	if (cl->state == HECI_CL_CONNECTED) {
+	if ((dev->dev_state == HECI_DEV_ENABLED) &&
+		(cl->state == HECI_CL_CONNECTED)) {
 		cl->state = HECI_CL_DISCONNECTING;
 		dev_dbg(&dev->pdev->dev, "disconnecting client host client = %d, ME client = %d\n",
 			cl->host_client_id, cl->me_client_id);
@@ -155,9 +153,7 @@ static int heci_release(struct inode *inode, struct file *file)
 	file->private_data = NULL;
 
 	/* disband and free all Tx and Rx client-level rings */
-	spin_lock_irqsave(&dev->cl_list_lock, flags);
 	heci_cl_free(cl);
-	spin_unlock_irqrestore(&dev->cl_list_lock, flags);
 
 	ISH_DBG_PRINT(KERN_ALERT "%s(): ---\n", __func__);
 	return rets;
diff --git a/drivers/misc/intel-ish/heci-hid-client.c b/drivers/misc/intel-ish/heci-hid-client.c
index ddd7ad9..3f4f97c 100644
--- a/drivers/misc/intel-ish/heci-hid-client.c
+++ b/drivers/misc/intel-ish/heci-hid-client.c
@@ -543,6 +543,13 @@ int     hid_heci_cl_remove(struct heci_cl_device *dev)
 	ISH_DBG_PRINT(KERN_ALERT "%s(): +++\n", __func__);
 	heci_hid_remove();
 	hid_heci_client_found = 0;
+
+	heci_cl_unlink(hid_heci_cl);
+	heci_cl_flush_queues(hid_heci_cl);
+
+	heci_cl_free(hid_heci_cl);
+
+	/* disband and free all Tx and Rx client-level rings */
 	hid_heci_cl = NULL;
 
 	for (i = 0; i < num_hid_devices ; ++i) {
diff --git a/drivers/misc/intel-ish/platform-config.h b/drivers/misc/intel-ish/platform-config.h
index 49c3261..71df6fb 100644
--- a/drivers/misc/intel-ish/platform-config.h
+++ b/drivers/misc/intel-ish/platform-config.h
@@ -17,7 +17,7 @@
 #define PLATFORM_CONFIG__H
 
 /* Build ID string */
-#define BUILD_ID        "236-dx-si-dma-impl-upstream-rc"
+#define BUILD_ID	"243-reset-flow-release-fix"
 
 #define	ISH_DEBUG	0
 #if ISH_DEBUG
-- 
1.9.1

